### PR TITLE
[ADD] auth_signup_recaptcha

### PR DESCRIPTION
**Pull request by bram1000**
Before this commit captcha was not enabled on signup and reset password requests. This meant that a lot of emails could be sent by automated means, causing issues for clients.

Task-3626220

### CODE REVIEW

**Comment by dansanti** [comment]
hello, in 16.0 ReCaptcha is in legacy mode  only, you should  use in legacy way 

- reset_password.js:

`odoo.define('aut_signup_recaptcha.reset_password', function (require) {
    'use strict';

    var publicWidget = require('web.public.widget');
    const { ReCaptcha } = require("google_recaptcha.ReCaptchaV3");

    publicWidget.registry.ResetPasswordCaptcha = publicWidget.Widget.extend({
        selector: '.oe_reset_password_form',
        events: {
            'submit': '_onSubmit',
        },
        init() {
            this._super(...arguments);
            this._recaptcha = new ReCaptcha();
            this.shouldCatch = true;
        },
        async willStart() {
            return this._recaptcha.loadLibs();
        },
        _onSubmit(event) {
            if(this.shouldCatch) {
                event.preventDefault()
                this._recaptcha.getToken('portal_signup').then(tokenCaptcha => {
                    this.$el.append(`<input name="recaptcha_token_response" type="hidden" value="${tokenCaptcha.token}"/>`);
                    this.shouldCatch = false;
                    this.$el.submit();
                })
            }
        },
    })`




- signup.js


`odoo.define('aut_signup_recaptcha.reset_password', function (require) {
    'use strict';

    var publicWidget = require('web.public.widget');
    const { ReCaptcha } = require("google_recaptcha.ReCaptchaV3");
    
    publicWidget.registry.SignupCaptcha = publicWidget.Widget.extend({
        selector: '.oe_signup_form',
        events: {
            'submit': '_onSubmit',
        },
        init() {
            this._super(...arguments);
            this._recaptcha = new ReCaptcha();
            this.shouldCatch = true;
        },
        async willStart() {
            return this._recaptcha.loadLibs();
        },
        _onSubmit(event) {
            if(this.shouldCatch) {
                event.preventDefault()
                this._recaptcha.getToken('portal_signup').then(tokenCaptcha => {
                    this.$el.append(`<input name="recaptcha_token_response" type="hidden" value="${tokenCaptcha.token}"/>`);
                    this.shouldCatch = false;
                    this.$el.submit();
                })
            }
        },
    })
});`

**Review by beledouxdenis — COMMENTED** [review]
Shouldn't we do something for login as well ?

**Comment by beledouxdenis** [comment]
> Yes the recaptcha modules are still in legacy in 16 but this doesn't mean new code should also use the old way ;-)

I kinda agree that for the sake of uniformity of the code the old way should

That being said that will fuck up the forward port as the code you will need to be converted to the new way.

To be honest I do not really care, I care more about the added value / security this feature is bringing than some nitpicking regarding the code formatting.

File: addons/auth_signup_recaptcha/static/src/signup.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+};
+
+publicWidget.registry.SignupCaptcha = publicWidget.Widget.extend({
+    ...CaptchaFunctionality,
+    selector: ".oe_signup_form",
+    token_name: "signup",
+});
+
+publicWidget.registry.ResetPasswordCaptcha = publicWidget.Widget.extend({
+    selector: ".oe_reset_password_form",
```
[the remark this answers is not in the document]
Comment (beledouxdenis) [reply addons/auth_signup_recaptcha/static/src/signup.js:41]: So did you test your dev now ? : D

**Comment by beledouxdenis** [comment]
> If we decide to then add captcha to yet another route, we will need another link module. 

Why ? We just add the URL of the route to the `CAPTCHA_ROUTES` inside `google_recaptcha`, we had the new selector to target the form where it should be applied, and that's it. No link module needed.

> And what about captchas where it's not so easy to add?

Why is it not easy to add ? Can you elaborate ?

**Comment by beledouxdenis** [comment]
> Well actually... we could look for DOM selectors even if the modules are not yet installed, but that would mean that the code would always run. This means the code could reference things that do not yet exist, but are only supposed to run when they do exist.
> 
> Plus if we go that route, we should move to 17.0/18.0 so I can immediately make it compatible with turnstile and then backport to 16 :')

Well I was actually wondering how we will do in 17.0 where there is both google_recaptcha and turnstile, whatever we choose the module link solution or the one I suggest.

Because it would be a shame to end with twice the same overrides, python and js ! in both modules
Like for instance for signup you will need
```py
@http.route()
    def web_auth_signup(self, *args, **kw):
        if request.httprequest.method == 'POST' and not request.env['ir.http']._verify_request_recaptcha_token('signup'):
            raise UserError(_("Suspicious activity detected by Google reCaptcha."))
        return super().web_auth_signup(*args, **kw)
```
and 
```js
publicWidget.registry.SignupCaptcha = publicWidget.Widget.extend({
    ...CaptchaFunctionality,
    selector: ".oe_signup_form",
    token_name: "signup",
});
```
in both `website_cr_turnstile` and `google_recaptcha`, as they do not depend on each other and their only common dependency is `base`.

And this is really not ideal.

**Comment by LuvForAirplanes** [comment]
What's the timeline on this be available in v18?

**Pull request closed without merging**